### DISCUSSION TITLE
Is it possible to make custom vehicle profiles?

### DISCUSSION THREAD

**Discussion by samvanemelen** [valhalla/valhalla#3853] [Q&A]
Is it possible to make custom vehicle profiles? I am working on routing for city logistics, but an issue I came across is that trucks are never routed through pedestrian zones. But pedestrian zones are often more complex with access times etc, I want to control access restrictions myself. I couldn't seem to find this online, but are there ways to create custom vehicle profiles so I can control which streets a vehicle can access?

I saw the [doc about costing options for the turn-by-turn api](https://github.com/valhalla/valhalla/blob/master/docs/api/turn-by-turn/api-reference.md#costing-options) that can handle custom speeds etc, I presume these will also work for the optimized-route? But I couldn't see an option to allow pedestrian streets.

Or is the best solution to manipulate the 'bicycle' profile to behave like a truck?

**Comment by nilsnolde** [comment]
Can you share which OSM tagging for pedestrian zones you mean? The `area=yes` ones or just ways tagged as `highway=pedestrian` with some other tags? If the former we have other problems on top of that, we can only route on the perimeter so far. The latter _might_ be possible.

The way I think about it is that no motorized vehicle should ever pass a `highway=pedestrian` unless tagged otherwise. So, by default it shouldn't be used for trucks, not even to off-load/do service. If in combination with `hgv=destination` or `access=delivery` (same for me, but maybe also includes cars), a truck could _start_ or _end_ a trip on that edge. Or if `hgv=yes` it could drive through (hopefully that's not a thing!).

The most common combination you'd like to cover is probably with `hgv=destination` or similar, for service trucks to deliver stuff to businesses. We can't (yet) have mode-specific `destination_only` edges, i.e. allow a car to traverse but not a truck, see https://github.com/valhalla/valhalla/issues/2563#issuecomment-1323641569. Once that's done, we could support that as well I think.

I'm not sure which conditional (access) restrictions support time in Valhalla, e.g. `destination_only` etc. Possibly there might be some missing, couldn't find out quickly. Maybe @kevinkreiser , @gknisely or @dnesbitt61 know off the top of their head.

**Reply by nilsnolde** [reply]
So, in summary, I don't think you'll need to look at coming up with a custom costing model. We could (and should) support that out-of-the-box IMO.

**Reply by nilsnolde** [reply]
What also will be harder (but might be, or is already, possible by duplicating access restrictions), is the scenario where both car & truck are allowed on an edge but only trucks are time-restricted, e.g. only allowed from 8am-5pm, but cars always allowed as destination-only.

**Reply by nilsnolde** [reply]
Finally, I'm not so sure if we already support the case where your destination is inside a `dest_only` "region", i.e. a `access=destination` edge which is only reachable by other `access=destination` edges. Probably supported in the bidirectional ones as it's easier to detect there.

**Reply by samvanemelen** [reply]
it is about this pedestrian zone (e.g. [way/51776661](https://www.openstreetmap.org/way/51776661)). I know in fact that this zone is accessible for trucks during certain hours. But it is indicated as highway=pedestrian. So no matter what, trucks are not allowed here according to the routing.

Valhalla's time restrictions are not on point yet, so I did some code myself to make sure the routes are allowed in the real world.

Agreed that by default trucks should never route through pedestrian zones, but as I understand it, the 'destination_only' thing would not work yet for pedestrian roads?

![image](https://user-images.githubusercontent.com/36103001/204586036-16211042-da9c-4609-883c-bf73680d7f1e.png)

**Reply by nilsnolde** [reply]
> I know in fact that this zone is accessible for trucks during certain hours [...] So no matter what, trucks are not allowed here according to the routing.

Restricted truck/car access is not reflected in the tagging of that way. You'd need to add smth like `hgv:conditional=destination@<time range(s)>` (whatever's actually the case) for routing engines to even be able in theory to know that. With the current tagging we will never allow a truck on a `pedestrian` way.

> Valhalla's time restrictions are not on point yet,

What do you mean exactly? I never really looked into the time-dependent access restriction logic in Valhalla, but it definitely helps which tags you think/know are not working with time-based restrictions, ideally with an example way's OSM ID so we can see the full tagging.

> but as I understand it, the 'destination_only' thing would not work yet for pedestrian roads?

Right, it doesn't at this point. My whole point was that we/I would like to support what you're thinking about. So it will be available in the future in Valhalla `master`. I know at least 3 clients who'd be very interested to have that supported. Still, at least I can't really promise anything time-wise (maybe someone else can..), only for sure will it happen within 2023. There's a few open issues that go in a similar direction where individual costing models could be improved, maybe we should create a new Github project and collect them..

If you can wait a little, it'd be worth it probably. If you're C++/routing-savvy and have the long breath to work on a PR yourself, we'd be happy to guide you a bit. It's not trivial though.

**Reply by samvanemelen** [reply]
Thank you @nilsnolde for the replies. I know that it is not reflected in the tags. According to the tags, trucks are never allowed here as you said. I understand that. I said that part because in the code I am writing, I am taking into account the actual time windows trucks are allowed in (that I know from knowing the city). What I am facing now, is how can I make the routing ignore the fact that according to the tags, trucks are never allowed in here. But as I understand it, even with tags about time conditional access, valhalla wouldn't use them at this point?

When I was discovering valhalla I was also curious to see if valhalla could optimize routes, not only by lowest cost, but also by time windows certain deliveries should be made. I can't seem to find it again, but I read somewhere that this feature was an ambition, but not developed yet. That is what I meant by "Valhallas time restrictions (for deliviers) are not on point yet"'. As reasoning to why I did these time checks myself, and just wanted to circumvent the "trucks not allowed in pedestrian streets" thing.

I would love to contribute to valhalla, because I am more than thankful to the features it makes available. But sadly I am not at the point where I would be able to make such changes to the core routing algorithms. I don't mind waiting for that.

Would using the bike profile (with changed speeds and other routing costs) be a way to circumvent these restrictions temporarily? Or are there other differences you can think of out of the top of your head that would result in different routes?

**Reply by nilsnolde** [reply]
> I am taking into account the actual time windows trucks are allowed in

I don't think you're able to do that outside of the routing engine/Valhalla. We need to know the exact time  when we encounter an edge with time conditions and that only a routing algorithm can do reliably. E.g. you're starting a route at 7 am a route which will take 2 hours and somewhere along the way are edges which are restricted from 8am-5pm, only the routing engine can tell if it's currently already past 8am or not. It's sort of edge effects, but important ones for full compliance.

**Comment by nilsnolde** [comment]
> I am taking into account the actual time windows trucks are allowed in (that I know from knowing the city).

I can only encourage you to add those to OSM, so everyone can use them properly. If you're unfamiliar with it, I'd try to contact your local OSM chapter if you have one or just ask in the forum.

> But as I understand it, even with tags about time conditional access, valhalla wouldn't use them at this point?

I'm not sure what exactly is supported on time-dependent restrictions currently. I'll dive into it at some point, but it'd take a bit of time, it's crossing quite a few modules inside Valhalla. Someone else surely knows better without having to look at the code.

> if valhalla could optimize routes, not only by lowest cost, but also by time windows certain deliveries should be made

Ah ok, those are _optimization_ restrictions, not routing restrictions. Valhalla can do a _very_ simple TSP optimization, and I doubt we'll ever want to do more than that. We do soon have a Chinese Postman implementation, but that's again very different territory and makes sense to include inside Valhalla. You're there crossing into the domain of VRP (vehicle routing problems) and currently the easiest and most feature-rich/performant solution is probably [VROOM](https://github.com/VROOM-Project/vroom), which supports time windows, skills, heterogeneous fleets and other goodies, AND works with Valhalla out-of-the-box (but not fully customizable yet, in worst case you'll have to send custom matrices to VROOM).

> Would using the bike profile (with changed speeds and other routing costs) be a way to circumvent these restrictions temporarily

Not sure what exact restrictions you'd like to circumvent temporarily tbh. I don't know of any meaningful time-based restrictions for bikes.
